assayist/client/query.py

```diff
@@ -15,6 +15,66 @@ def set_connection(neo4j_url):
     neomodel.db.set_connection(neo4j_url)
 
 
+def get_container_by_component(component_name, component_version):
+    """Query for builds by component name and version.
+
+    Finds builds that produced container images which include or embed a component with the
+    specified name and version, and any version preceding that version.
+
+    :param str component_name: name of the component to query for
+    :param str component_version: version of the component to query for
+    :return: list of builds IDs
+    :rtype: list
+    """
+    query = """
+        // Find a component and a sourcelocation with the specified name and version. These can
+        // be emdedded in other sourcelocation that are used as a source for a build or directly
+        // used for a build.
+        MATCH (c:Component {{canonical_name: '{name}'}}) <-[:SOURCE_FOR]-
```

Comment:
Since this is a `MATCH` and not an `OPTIONAL MATCH`, the whole query will return nothing if none of the nodes in `target_sl` have a build associated with it. Based on the supersedes optional match below, I don't think that's what you want. I suggest splitting this up in two queries (untested):

```cypher
 MATCH (c:Component {{canonical_name: '{name}'}}) <-[:SOURCE_FOR]-version_sl:SourceLocation {{canonical_version: '{version}'}})<-[:UPSTREAM|EMBEDS*0..]- (target_sl:SourceLocation)

OPTIONAL MATCH (target_sl)<-[:BUILT_FROM]- (direct_builds:Build)
```